### PR TITLE
Add left and right distributive properties

### PR DESCRIPTION
**Pull request by gwils**
What do you reckon about this? Any recommendations on cleaning it up or other suggestions?

### CODE REVIEW

**Comment by gwils** [comment]
Don't merge this yet. The tests fail even though Travis is happy.

**Comment by bitemyapp** [comment]
I merged the other branch, I'll see if I can bounce this in Travis.

**Comment by bitemyapp** [comment]
Pop goes the weasel. I'll look at the code now.

File: hedgehog-checkers/src/Hedgehog/Checkers/Properties.hs [hunk 1]
```diff
@@ -71,3 +77,28 @@ commutativity f gena = do
...
 --     x < y || x == y || x > y
+
+leftDistributive :: ( Ord a
+                    , Eq (f b)
+                    , Show (f a)
+                    , Show (f b)
+                    )
+                 => (forall p q. (p -> q) -> f p -> f q) -> (forall z. f z -> f z -> f z) -> Gen (f a) -> Gen a -> Gen b -> PropertyT IO ()
+leftDistributive x y genFa genA genB = do
+  f <- ordFuncWtf genA genB
```
Comment (bitemyapp) [inline hedgehog-checkers/src/Hedgehog/Checkers/Properties.hs:88]: You don't need to change anything, but I really dislike `ordFuncWtf`. It's so slow if you want more than minimal coverage.

**Comment by bitemyapp** [comment]
This looks great to me, particularly given you factored out left and right distributivity.

@jystic I have a question, is the right thing for gathering together some logically related properties to monadically combine then into a single `PropertyT ...`? Thinking of things like `functor` which has two laws. I tried organizing them into `Group`s but it was a pain to pull them together in a test suite from that.

Leaving everything `PropertyT` has been easier so far, with any factoring out being to `-> Bool` functions that get pulled into properties.

File: hedgehog-checkers/src/Hedgehog/Checkers/Properties.hs (hunk 1 above)
Comment (gwils) [reply hedgehog-checkers/src/Hedgehog/Checkers/Properties.hs:88]: I agree, but I don't know what more can be done.
Speaking of that function, I wouldn't mind renaming that module and its contents.

Comment (bitemyapp) [reply hedgehog-checkers/src/Hedgehog/Checkers/Properties.hs:88]: >Speaking of that function, I wouldn't mind renaming that module and its contents.

Sure, as long as the new nomenclature retains the useful aposematism.

**Comment by gwils** [comment]
Closing this, because I've got a better branch. It depends on some small refactorings I made in #10, so let's resolve that first and come back to distributivity :)

**Pull request closed without merging**